### DISCUSSION TITLE
Extending Typescript JobAttributesData with custom properties error?

### DISCUSSION THREAD

**Discussion by uncvrd** [agenda/agenda#1316] [Q&A]
Hi all - am I missing something completely obvious? I'd like to add custom properties to my data object to access while running a job. In the typescript typings it shows: 

```
declare class Job<T extends JobAttributesData = JobAttributesData>{}
```

Which means that I should be able to define an interface that extends `JobAttributesData` like:

```
interface Test extends JobAttributesData {
    prop: string;
}
```

Then use in my job definition callback like so:

```
this.client.define("test", async (job: Job<Test>) => {
    console.log(job.attrs.data.prop)
    return;
})
```

However, typescript shows a syntax error on the `async (job: Job<Test>)` stating:

```
Argument of type '(job: Job<Test>) => Promise<void>' is not assignable to parameter of type 'DefineOptions | Processor'.
  Type '(job: Job<Test>) => Promise<void>' is not assignable to type '(job: Job<JobAttributesData>) => Promise<void>'.
    Types of parameters 'job' and 'job' are incompatible.
      Type 'Job<JobAttributesData>' is not assignable to type 'Job<Test>'.
        Property 'prop' is missing in type 'JobAttributesData' but required in type 'Test'.ts(2345)
```

What can I do to resolve this? Is my understanding of extending incorrect? Thanks!

```
"agenda": "^4.1.3",
"@types/agenda": "^4.1.0",
```

EDIT:

Is this because I am attempting to extend from `JobAttributesData` which as the signature:

```
export interface JobAttributesData {
    [key: string]: any;
}
```

So when I extend from that, it has no way of identifying if I am using a dynamic key name or not? I see that I can in-fact, work around this by defining my `Test` interface as:

```
interface Test extends JobAttributesData {
    prop?: string; // note the ?, which makes this key optional
}
```

**Comment by pavel-lens** [comment]
👍 I have the same problem after migrating from Agenda 3 to Agenda 4. I'm extending the `JobAttributesData` interface, but the TS compiler is complaining. Can you please elaborate @boredland ?

**Reply by boredland** [reply]
sorry, I didn't use agenda for a long time: how am I connected to this?

**Reply by pavel-lens** [reply]
Oh, sorry then. I looked at who edited the last the types and it was you. I will look for a core contributor.

**Reply by boredland** [reply]
I see, I added the types here! Did not remember that. I think I understand the problem.

**Reply by pavel-lens** [reply]
I created a PR to resolve the issue (for everyone's future reference)
https://github.com/agenda/agenda/pull/1463

**Reply by boredland** [reply]
ah, ya, that's the issue I thought of too! Very nice!

**Reply by boredland** [reply]
I think back [then](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51150/files#diff-ce4d15ed47e1b39240f46a09c40196b89a89f94f28e51ebcdae2a27ec33c9fa8) I tried to make the old DT-Types work inlined, nothing more, nothing less. But apparently they were not that great to begin with.
